### PR TITLE
feat: Implement KeyDictBuilder for the merge tree memtable

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR implements the KeyDictBuilder for the merge tree memtable
- Each builder has fixed capacity
- It interns key and returns an index to get the key from the builder again
- It can scan keys in sorted order from the builder
- It can build a KeyDict by invoking `finish()`
  - It will replace pk index with a new value so the pk index has the same order the key

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

## Refer to a related PR or issue link (optional)
- #2804

### CODE REVIEW

File: src/mito2/src/memtable/merge_tree/dict.rs [hunk 1]
```diff
@@ -14,15 +14,400 @@
...
+
+    use super::*;
+
+    fn prepare_input_keys(num_keys: usize) -> Vec<Vec<u8>> {
+        let prefix = ["a", "b", "c", "d", "e", "f"];
+        let mut rng = rand::thread_rng();
+        let mut keys = Vec::with_capacity(num_keys);
+        for i in 0..num_keys {
+            let prefix_idx = rng.gen_range(0..prefix.len());
+            // We don't need to decode the priamry key in index's test so we format the string
```
Comment (fengjiachun) [inline src/mito2/src/memtable/merge_tree/dict.rs:354]:
```suggestion
            // We don't need to decode the primary key in index's test so we format the string
```

File: src/mito2/src/memtable/merge_tree/dict.rs [hunk 2]
```diff
@@ -14,15 +14,400 @@
...
-pub struct KeyBuffer {}
+impl KeyDictBuilder {
+    /// Creates a new builder that can hold up to `capacity` keys.
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            capacity,
+            num_keys: 0,
+            pk_to_index: BTreeMap::new(),
+            key_buffer: KeyBuffer::new(MAX_KEYS_PER_BLOCK.into()),
+            dict_blocks: Vec::new(),
```
Comment (fengjiachun) [inline src/mito2/src/memtable/merge_tree/dict.rs:55]: Non-blocking： `dict_blocks: Vec::with_capacity(capacity / MAX_KEYS_PER_BLOCK + 1),`

File: src/mito2/src/memtable/merge_tree/dict.rs [hunk 3]
```diff
@@ -14,15 +14,400 @@
...
+    }
+
+    /// Returns current [PkIndex] of the key.
+    pub fn current_pk_index(&self) -> PkIndex {
+        assert!(self.is_valid());
+        self.sorted_pk_indices[self.offset]
+    }
+
+    /// Advances the reader.
+    pub fn next(&mut self) {
```
Comment (fengjiachun) [inline src/mito2/src/memtable/merge_tree/dict.rs:188]: Question: Can the `next` method directly return `current_pk_index` and `current_key`? Then remove the two methods `current_key` and `current_pk_index`.

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/mito2/src/memtable/merge_tree/dict.rs [hunk 4]
```diff
@@ -14,15 +14,400 @@
...
+    }
+
+    /// Adds the key to the builder and returns its index if the builder is not full.
+    ///
+    /// Returns `None` if the builder is full.
+    pub fn try_insert_key(
+        &mut self,
+        key: &[u8],
+        metrics: &mut WriteMetrics,
+    ) -> Result<Option<PkIndex>> {
```
Comment (v0y4g3r) [inline src/mito2/src/memtable/merge_tree/dict.rs:72]: Looks like this method does not have to return a result.

File: src/mito2/src/memtable/merge_tree/dict.rs (hunk 3 above)
Comment (v0y4g3r) [reply src/mito2/src/memtable/merge_tree/dict.rs:188]: We need to manually advance the reader and read the current pk_index multiple times.

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**